src/java/org/apache/cassandra/index/sai/plan/Expression.java

```diff
@@ -0,0 +1,380 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.plan;
+
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.cql3.Operator;
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.analyzer.AbstractAnalyzer;
+import org.apache.cassandra.index.sai.utils.TypeUtil;
+
+public class Expression
+{
+    private static final Logger logger = LoggerFactory.getLogger(Expression.class);
+
+    public enum Op
+    {
+        EQ, RANGE, CONTAINS_KEY, CONTAINS_VALUE, IN;
+
+        public static Op valueOf(Operator operator)
+        {
+            switch (operator)
+            {
+                case EQ:
+                    return EQ;
+
+                case CONTAINS:
+                    return CONTAINS_VALUE; // non-frozen map: value contains term;
+
+                case CONTAINS_KEY:
+                    return CONTAINS_KEY; // non-frozen map: value contains key term;
+
+                case LT:
+                case GT:
+                case LTE:
+                case GTE:
+                    return RANGE;
+
+                case IN:
+                    return IN;
+
+                default:
+                    return null;
+            }
+        }
+
+        public boolean isEquality()
+        {
+            return this == EQ || this == CONTAINS_KEY || this == CONTAINS_VALUE;
+        }
+
+        public boolean isEqualityOrRange()
+        {
+            return isEquality() || this == RANGE;
+        }
+    }
+
+    public final AbstractAnalyzer.AnalyzerFactory analyzerFactory;
+
+    public final IndexContext context;
+    public final AbstractType<?> validator;
+
+    @VisibleForTesting
+    protected Op operation;
+
+    public Bound lower, upper;
+    public boolean upperInclusive, lowerInclusive;
```

Comment:
Both `lower` and `upper` have an `inclusive` attribute. How are they different from `upperInclusive` and `lowerInclusive`? I'd add some JavaDoc to the attributes about this.

Reply:
I think these were leftover from a previous implementation. The `inclusive` attribute on `Bound` is already being used in a number of places. I'll remove `upperInclusive` and `lowerInclusive`.